hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java

```diff
@@ -93,16 +93,13 @@ public WritableECContainerProvider(ConfigurationSource conf,
   public ContainerInfo getContainer(final long size,
       ECReplicationConfig repConfig, String owner, ExcludeList excludeList)
       throws IOException, TimeoutException {
-    // Bound this at a minimum of 1 byte in case a request is made for a very
-    // small size, which when divided by EC DataNum is zero.
-    long requiredSpace = Math.max(1, size / repConfig.getData());
     synchronized (this) {
       int openPipelineCount = pipelineManager.getPipelineCount(repConfig,
           Pipeline.PipelineState.OPEN);
       if (openPipelineCount < providerConfig.getMinimumPipelines()) {
         try {
           return allocateContainer(
-              repConfig, requiredSpace, owner, excludeList);
+              repConfig, size, owner, excludeList);
```

Comment:
I'm not sure this is correct. I remember looking into what this "size" is that is passed into this method. In the Java doc, I wrote

```
   * @param size The max size of block in bytes which will be written. This
   *             comes from Ozone Manager and will be the block size configured
   *             for the cluster. The client cannot pass any arbitrary value
   *             from this setting.
```

So if the block size gets passed in (256MB) and we track an EC container as having a 5GB limit for each replica index, then the space we need in the container is 256 / dataNum.

Now if the size at OM is divided by dataNum already and then passed down to SCM, then we don't need to divide it again, but I don't think it is, at least it did not used to be.

Reply:
The `size` passed to `getContainer` is scmBlockSize (256 MB), and it's called in allocateBlock.

https://github.com/apache/ozone/blob/a02c6df497b2a636c7a81c26e53f39daa7958841/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableContainerFactory.java#L54-L61

If this is expected, we should revisit the line 148 in the following code.

https://github.com/apache/ozone/blob/9583e4733155f418fe4ceda6c26f2117eec4c137/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java#L138-L150

Reply:
> https://github.com/apache/ozone/pull/3194#discussion_r826944617
> For EC we are allocating in "ec block" chunks, which is always 256 * 5 (rs-3-2). We are not allocating in 256MB chunks. The code allocatedLocationInfoList.size() should be returning the number of EC Block groups allocated.

If I understand correctly, we should either pass the `requested data size` to `allocateBlock()` and devide it inside `WritableECContainerProvider`,
or pass `requested data size / numData` to `allocateBlock()` and use the raw value in `WritableECContainerProvider`.

Based on the definition of `OMAllocateBlockRequest`, seems the first approach is the correct one?

Reply:
The cluster blocksize is 256MB, but when we ask for an EC block, we are really asking for 256MB * dataNum.

Right now, we only pass 256MB down to SCM and do not adjust it for the dataNum. Then is SCM, right now we are dividing it by dataNum.

Now I look at this again, I think you are correct in what you said before. We should not divide this in SCM, as its already the "single replica block size" that is being passed in.

Reply:
I think our two messages crossed over.

The issue we have, is that the container size limit is 5GB for both Ratis and EC. But for EC, each container in the group has a capacity of 5GB. When we are searching for a container with free space, we use just the 5GB limit.

Right now, the OM code just passes the cluster block size down to SCM to request a container. Also now, we are dividing that by dataNum. However that is not correct, which you pointed out earlier.

The EC logical block is 256MB * dataNum, but since we don't scale up the EC container size limit, we should just use blockSize without multiplying or dividing it.